mathics/builtin/patterns.py

```diff
@@ -588,9 +621,14 @@ def yield_match(vars_2, rest):
             else:
                 yield_func(vars_2, None)
 
-        self.pattern.match(yield_match, expression, vars, evaluation)
+        pattern_context = pattern_context.copy()
+        pattern_context["yield_func"] = yield_match
```

Comment:
I don't understand why the `copy` and `pattern_context`  now is needed when it wasn't before. Was the previous code wrong?  Please enlighten me - thanks.

Reply:
I agree that using a fixed parameter is less prone to error, and allows for automated error checking.  

As for the efficiency part, unless this very much slows things down, let's revisit this. It may be that some future way of handling patterns will make all of this moot. 

It would be good to note somewhere in the code that a copy is needed because parameters can sometimes get altered, maybe with an example.

Reply:
> As for the efficiency part, unless this very much slows things down, let's revisit this. It may be that some future way of handling patterns will make all of this moot.
> 

Also, using copy() avoids explicitly rebuilding each entry in the dict, which makes the code shorter.

> It would be good to note somewhere in the code that a copy is needed because parameters can sometimes get altered, maybe with an example.

I thought I had put a comment in `BasePattern.match`, but it seems it was lost in one of my failed attempts that this works. Also, I have to adjust its docstring. Maybe in the next round.

Reply:
It looks like `match_numericq()`, `match_real_numberq()` and other `match_xxx()` could be DRY'd by having a common function that takes a `yield_match` function in as a parameter.

The other good thing about doing something like this is that it makes it clearer what the _difference_ is between these routiens.

Reply:
@rocky, as I mentioned before, there are a lot of things to do just to make the code more readable and easy to understand and debug, but a larger PR would make it harder to review. If you agree, in this round we can leave a TODO comment and face it in the following round.

Reply:
No problem. But since I am looking at this, I want to make sure I note what I see in this pass.

I am now done making a pass over the PR. The final changes suggested are in the `more_patterns_more` branch.